compiler/passes/src/write_transforming/mod.rs

```diff
@@ -0,0 +1,53 @@
+// Copyright (C) 2019-2025 Provable Inc.
+// This file is part of the Leo library.
+
+// The Leo library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The Leo library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::Pass;
+
+use leo_ast::ProgramReconstructor as _;
+
+use leo_errors::Result;
+
+mod expression;
+
+mod program;
+
+mod statement;
+
+mod visitor;
+use visitor::*;
+
+/// A pass to rewrite assignments to array accesses and struct accesses.
+///
+/// The pass requires that the AST is in SSA form and that tuples have been
+/// destructured. Since the pass will create new assignments, `SsaForming` must
+/// be run again afterwards.
+pub struct WriteTransforming;
```

Comment:
Including a sketch of the before and after ASTs would be helpful in giving the reader an idea of the transformation.